engine/src/flutter/lib/ui/painting/image_decoder_impeller.cc

```diff
@@ -394,8 +394,17 @@ ImageDecoderImpeller::UnsafeUploadTextureToPrivate(
     result_texture = std::move(resize_texture);
   }
   blit_pass->EncodeCommands();
-
-  if (!context->GetCommandQueue()->Submit({command_buffer}).ok()) {
+  if (!context->GetCommandQueue()
+           ->Submit(
+               {command_buffer},
+               [](impeller::CommandBuffer::Status status) {
+                 if (status == impeller::CommandBuffer::Status::kError) {
+                   FML_LOG(ERROR)
+                       << "GPU Error submitting image decoding command buffer.";
+                 }
+               },
+               /*block_on_schedule=*/true)
```

Comment:
This is going to be inefficient, needlessly synchronizing the cpu and the gpu in most cases.

I think the better way to implement this would be for `ImageDecoderImpeller` to maintain a receipt from submitting the command buffer.  It will add a `fml::SyncSwitch::Observer` to the `fml::SyncSwitch`, when that notifies the `ImageDecoderImpeller` that we are losing access to the GPU it uses the receipt to get access to the buffer and call `waitUntilScheduled`.

Reply:
that is not what block_on_scheduled does. in fact, scheduling has nothing to do with gpu usage at all (that would be waitUntilCompleted). All it blocks on is the queue acknowledging the buffer, which is not synchronous unlike GLES/Vulkan.

Reply:
There is an asynchronous operation happening which `waitUntilScheduled` is causing the calling thread to block on.  The approach I'm proposing will only block the thread in the case of  `applicationDidEnterBackground:` which is my understanding of what Apple is recommending.

Reply:
Maybe not.  I'm not quite sure why we would prefer it though when we know of an alternative.  Plus the alternative doesn't make everyone that submits a command buffer learn what that flag means when it's only relevant on one platform, in one usage.

Reply:
Plus we're dealing with a bit of a black box with the driver.  I'd say it's preferred we follow Apple's recommendation.

Reply:
...and I'll add that we don't know if it fixes it for real or not. If we want to engineer a better solution if it works, I'm down for that. But if we end up back at square one because it doesn't fix it I plan to revert this.

Reply:
From offline chat: Calling waitForScheduled when we don't have access to the gpu will cause deadlock, but this is happening only inside of a syncswitch block to guarantee we have the GPU.